### ISSUE TITLE
Proposal: new option for no-unmodified-loop-condition: mutators

### ISSUE THREAD

**Issue by mnpenner** [eslint/eslint#8021]
**What rule do you want to change?**

[`no-unmodified-loop-condition`](http://eslint.org/docs/rules/no-unmodified-loop-condition)

**Does this change cause the rule to produce more or fewer warnings?**

Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option: `mutators`, accepts a regex like `/^set/` which will be matched against methods called on the loop conditionals. If one is found, it shall be assumed that the conditional was modified.

**Please provide some example code that this change will affect:**

```js
while (start <= end) { 
      dates.push(new Date(start))
      start.setDate(start.getDate() + 1)
}
```

**What does the rule currently do for this code?**

Produces an error:

> 'start' is not modified in this loop

**What will the rule do after it's changed?**

No error. Specifically, if you add the option `mutators: /^set/` as described above, then eslint should check all methods invoked on `start` and `end` within the loop body. In this case it would match on `setDate` so it should be assumed that `start` was modified, thereby alleviating the warning.

**Comment by feross** [comment]
I would like to ensure that *any* reference to the condition variables is treated as a modification. In this proposal, I assume I could specify the option: `mutators: /.*/` Is that correct?

**Comment by mnpenner** [comment]
@feross If by "any reference" you mean `conditionVar.anything()`, yes. My proposal doesn't cover merely using `conditionVar` without invoking any methods on it. For example, this would still generate a warning:

```js
while (start <= end) { 
    if(start) {
         console.log(start);
    }
}
```

We've used `start` in the body, but haven't modified in any way (either via `start = ...` or `start.xxx()`).

**Comment by feross** [comment]
@mnpenner Got it. Thanks for the clarification. That's still an improvement. If this proposal is implemented, we would use it in `standard`.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
